rhizome/lib/vm_setup.rb

```diff
@@ -67,13 +67,42 @@ def purge
     FileUtils.rm_f(vp.dnsmasq_service)
     r "systemctl daemon-reload"
 
+    purge_storage
+    r "umount #{vp.q_hugepages}"
+
     begin
       r "deluser --remove-home #{q_vm}"
     rescue CommandFail => ex
       raise unless /The user `.*' does not exist./.match?(ex.stderr)
     end
   end
 
+  def purge_storage
+    params = JSON.parse(File.read(vp.prep_json))
+    params["storage_volumes"].each { |disk|
+      device_id = disk["device_id"]
+      index = disk["disk_index"]
+
+      vhost_controller = Spdk.vhost_controller(@vm_name, index)
+
+      r "#{Spdk.rpc_py} vhost_delete_controller #{vhost_controller.shellescape}"
+
+      if disk["encrypted"]
+        q_keyname = "#{device_id}_key".shellescape
+        q_aio_bdev = "#{device_id}_aio".shellescape
+        r "#{Spdk.rpc_py} bdev_crypto_delete #{device_id.shellescape}"
+        r "#{Spdk.rpc_py} bdev_aio_delete #{q_aio_bdev}"
+        r "#{Spdk.rpc_py} accel_crypto_key_destroy -n #{q_keyname}"
+      else
+        r "#{Spdk.rpc_py} bdev_aio_delete #{device_id.shellescape}"
+      end
+
+      FileUtils.rm_f(Spdk.vhost_sock(vhost_controller))
```

Comment:
suggest not using "_f" variants with anything that could be construed as sensitive, because it's very relaxed about any kind of error that can show up with deletion, e.g. EACCESS will leave the file or directory intact and not error. Instead, don't use force and rescue Errno::ENOENT and continue with this as a success condition. This may be a worthy helper procedure of our own design.